appendices/VK_EXT_custom_border_color.txt

```diff
@@ -0,0 +1,89 @@
+include::{generated}/meta/VK_EXT_custom_border_color.txt[]
+
+*Last Modified Date*::
+    2019-10-14
+*IP Status*::
+    No known IP claims.
+*Contributors*::
+  - Joshua Ashton
+  - Liam Middlebrook, Nvidia
+
+This extension provides cross-vendor functionality to specify a custom border color
+for use when the sampler address mode ename:VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER is used.
+
+To create a sampler which uses a custom border color
+slink:VkSamplerCreateInfo::pname:borderColor must be set to ename:VK_BORDER_COLOR_CUSTOM_EXT.
+
+When ename:VK_BORDER_COLOR_CUSTOM_EXT is supplied, applications must provide a
+slink:VkSamplerCustomBorderColorCreateInfoEXT in the pNext chain for
+slink:VkSamplerCreateInfo.
+
+For implementations with a limited number of border colors, they should be reference counted
+when the sampler is created/destroyed. This means that freeing the last sampler that used
+a custom border color will also free up the slot for another border color to be used.
+
+The implementation should maintain the guarantee that flink:vkCreateSampler and
+flink:vkDestroySampler is thread-safe when a custom border color is used.
+
+=== New Enum Constants
+
+  * Extending elink:VkStructureType:
+  ** ename:VK_STRUCTURE_TYPE_SAMPLER_CUSTOM_BORDER_COLOR_CREATE_INFO_EXT
+  ** ename:VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_CUSTOM_BORDER_COLOR_PROPERTIES_EXT
+  ** ename:VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_CUSTOM_BORDER_COLOR_FEATURES_EXT
+
+  * Extending elink:VkBorderColor:
+  ** ename:VK_BORDER_COLOR_CUSTOM_EXT
+
+=== New Structures
+
+  * slink:VkSamplerCustomBorderColorCreateInfoEXT
+  * slink:VkPhysicalDeviceCustomBorderColorPropertiesEXT
+  * slink:VkPhysicalDeviceCustomBorderColorFeaturesEXT
+
+=== New Functions
+
+=== Issues
+
+1) Should VkClearColorValue be used for the border color value, or should we
+have our own struct/union?
+Do we need to specify the type of the input values for the components? This is
+more of a concern if VkClearColorValue is used here because it provides a union
+of float,int,uint types.
+
+OPEN
+
+2) For hardware which supports a limited number of border colors what happens if
+that number is exceed? Should this be handled by the driver unbeknownst to the
+application. In Revision 1 we had solved this issue using a new Object type,
+however that may have lead to additional system resource consumption which would
+otherwise not be required.
+
+OPEN
```

Comment:
The maxUniqueCustomBorderColors seems like the right thing to do, just need to be careful to specify precisely how/when things are counted.

Reply:
Is it really the case that each unique border color only counts once against this limit? I could imagine hardware where the border color depends on things like the swizzle or format or sRGBness. Swizzle and format are probably OK because implementations today would have a problem with runtime mix/match of images and samplers if they had this dependency, but sRGB might be a real lurking issue (not a problem today because 0/1 map to 0/1 in sRGB).

Reply:
I believe if we consider AMD and Intel then the UINT/FLOAT/INT-ness and Value should be enough. For vendors other than that, I have no idea.

Maybe leaving uniqueness open to the implementation and adding a function to query whether a color is unique would be an option, but IDK if that's really worth it, or if anyone would use it.